### PR TITLE
Feedback disambiguation

### PR DESCRIPTION
**Pull request by bwaresiak**
Task/Issue URL: https://app.asana.com/0/856498667320406/1109166030178657
Tech Design URL:
CC:

**Description**:
Updated feedback flow with categories and subcategories to disambiguate the feedback.

Q: When are Pixels fired?
A: Pixels are fired for each "submit feedback" action, which is:
 - For positive flow: 'rate app', 'no thanks' (on positive feedback screen) and 'submit feedback' (on feedback form).
 - For positive flow: Every time user presses 'submit feedback'.

Q: Where we can see feedback?
A: Asana [Mobile/AppFeedback](https://app.asana.com/0/1114857721287938/list) project and [Grafana](https://grafana.duckduckgo.com/d/YDs3FwCmz/ios-app-feedback)

Q: What about old feedback form?
A: We leave it as it is for now.

**Steps to test this PR**:
UI-related corner cases:
1. iPhone SE in landscape mode.
2. iPad of regular size.
3. iPad Pro 12,9 inch.

For all of the above, observer if keyboard&scrolling works properly on Feedback Form (in both regular and Negative Feedback->Broken website flow).

Feedback Flow - Category & Subcategory
1. Navigate to settings -> send feedback.
2. Select negative flow.
3. Select feedback category other than 'None of these' and 'Website issues'.
4. Select feedback subcategory other than 'None of these'
5. Ensure feedback form is populated with respective choices.

Feedback Flow - Category & Other
1. Navigate to settings -> send feedback.
2. Select negative flow.
3. Select feedback category other than 'None of these' and 'Website issues'.
4. Select feedback subcategory 'None of these'
5. Ensure feedback form is populated with respective choices (subtitle should be of generic type).

Feedback Flow - Website issues
1. Navigate to settings -> send feedback.
2. Select negative flow.
3. Select feedback category 'Website issues'.
4. Ensure feedback form has additional URL field and labels are related to Website issues.

Feedback Flow - Other issues
1. Navigate to settings -> send feedback.
2. Select negative flow.
3. Select feedback category 'None of these'
4. Ensure feedback form has generic title/subtitle.

Feedback Flow - Positive flow -> Rate app
1. Navigate to settings -> send feedback.
2. Select positive flow.
3. Tap 'rate the app' - AppStore should be opened.

Feedback Flow - Positive flow -> Share Details
1. Navigate to settings -> send feedback.
2. Select positive flow.
3. Tap 'Share Details' - Feedback form should be presented with correct image and title/subtitle/placeholder.

Feedback Flow - Positive flow -> No thanks
1. Navigate to settings -> send feedback.
2. Select positive flow.
3. Tap 'No thanks' - view is dismissed.

Report broken site
1. Navigate to some webpage.
2. Options -> Report broken site
3. See that this form works as expected.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo/FeedbackViewController.swift [hunk 1]
```diff
@@ -2,7 +2,7 @@
 //  FeedbackViewController.swift
 //  DuckDuckGo
 //
-//  Copyright © 2018 DuckDuckGo. All rights reserved.
+//  Copyright © 2019 DuckDuckGo. All rights reserved.
```
Comment (bwaresiak) [inline DuckDuckGo/FeedbackViewController.swift:5]: "Old" FeedbackVC became SiteFeedbackVC and this is actually a new file.

File: DuckDuckGo/FeedbackFormViewController.swift [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+        guard let keyboardFrame = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect else { return }
+        let keyboardSize = keyboardFrame.size
+        let contentInsets = UIEdgeInsets(top: 0.0, left: 0.0, bottom: keyboardSize.height, right: 0.0)
+        scrollView.contentInset = contentInsets
+        scrollView.scrollIndicatorInsets = contentInsets
+        messageViewHeight.constant = Constants.defaultMessageViewHeight
+        
+        let interactionRect = interactionAreaForEditingMode()
+        
+        if let rect = scrollView.desiredVisibleRect(forInteractionArea: interactionRect,
```
Comment (brindy) [inline DuckDuckGo/FeedbackFormViewController.swift]: guard and return would be a little cleaner

File: DuckDuckGo/FeedbackViewController.swift [hunk 3]
```diff
@@ -17,195 +17,78 @@
...
+        supplementaryText.setAttributedTextString(UserText.feedbackStartSupplementary)
+        footerText.setAttributedTextString(UserText.feedbackStartFooter)
     }
-
-    private func refreshButton() {
-        submitButton.isEnabled = feedbackModel.canSubmit()
-        submitButton.backgroundColor = submitButton.isEnabled ? UIColor.cornflowerBlue : UIColor.mercury
+    
+    private func configureButtons() {
+        postitiveFeedbackButton.layer.cornerRadius = 8
```
Comment (brindy) [inline DuckDuckGo/FeedbackViewController.swift]: can this duplication be reduced?

File: DuckDuckGo/FeedbackFormViewController.swift [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        loadViewIfNeeded()
+        
+        headerImage.image = UIImage(named: "sadFace")
+        self.headerText.text = FeedbackPresenter.title(for: category)
+        
+        switch type {
+        case .regular:
+            hideWebsiteField()
+            
+            if let subcategory = feedbackModel.subcategory {
```
Comment (brindy) [inline DuckDuckGo/FeedbackFormViewController.swift:111]: maybe extract this to a private function to make this section a bit more readable?

File: Core/Pixel.swift [hunk 5]
```diff
@@ -113,14 +116,26 @@ public class Pixel {
...
                             withHeaders headers: HTTPHeaders = APIHeaders().defaultHeaders,
                             onComplete: @escaping (Error?) -> Void = {_ in }) {
+        fire(rawPixel: pixel.rawValue,
+             forDeviceType: deviceType,
+             withAdditionalParameters: params,
+             withHeaders: headers,
+             onComplete: onComplete)
+    }
+    
+    public static func fire(rawPixel: String,
```
Comment (brindy) [inline Core/Pixel.swift]: Thinking more about this, I'd rather keep it type-safe:

* It's safer
* It's cleaner
* It'll be easier to change (refactoring pixel usage)
* It'll keep defining new ones consistent

I'd be happy with the fire function taking a `subCategories` parameter which defaults to an array of empty strings.  Then to create the raw pixel name in the fire function you can just do something like:

`let rawPixel = ([pixel.rawValue] + subPixels).joined(separator: "_")`

File: DuckDuckGo/PositiveFeedbackViewController.swift [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    private func configureButtons() {
+        leaveFeedbackButton.layer.borderWidth = 1
+    }
+    
+    @IBAction func rateAppButtonPressed() {
+        FeedbackSubmitter().firePositiveSentimentPixel()
+        
+        let urlStr = "itms-apps://itunes.apple.com/us/app/duckduckgo-privacy-browser/id663592361?action=write-review"
+
+        UIApplication.shared.open(URL(string: urlStr)!)
```
Comment (brindy) [inline DuckDuckGo/PositiveFeedbackViewController.swift:58]: should this also close the feedback form?

**Review by brindy — COMMENTED** [review]
Still testing - few minor pieces of feedback so far...

**Comment by brindy** [comment]
Generally looks good, subject to the feedback already given.  There were some changes came up in O-A - you'll see those tomorrow.  Also a conflict to resolve.  Should be a quick re-review when you need it. 👍

File: DuckDuckGo/PositiveFeedbackViewController.swift (hunk 6 above)
Comment (bwaresiak) [reply DuckDuckGo/PositiveFeedbackViewController.swift:58]: Yes, my bad :)

**Comment by bwaresiak** [comment]
@brindy Thanks for the feedback! :) This should be ready for re-review.

**Review by brindy — APPROVED** [review]
LGTM!

**Comment by bwaresiak** [comment]
Thanks!
